davinci/CONCEPT.md

```diff
@@ -0,0 +1,136 @@
+CollectorFactory.register("PASSWORD", ::PasswordCollector)
+
+CollectorFactory.register("SUBMIT_BUTTON", ::SubmitCollector)
+
+// Allow to map multiple Field Type to the same Collector
+CollectorFactory.register("FLOW_BUTTON", ::FlowCollector)
+CollectorFactory.register("FLOW_LINK", ::FlowCollector)
+```
+
+## How Collector is created & initialized
+
+DaVinci Response JSON:
+
+```json
+{
+  "form": {
+    "components": {
+      "fields": [
+        {
+          "type": "TEXT",
+          "key": "user.username",
+          "label": "Username",
+          "required": true,
+          "validation": {
+            "regex": "^[^@]+@[^@]+\\.[^@]+$",
+            "errorMessage": "Must be alphanumeric"
+          }
+        },
+        {
+          "type": "PASSWORD",
+          "key": "password",
+          "label": "Password",
+          "required": true
+        },
+        ...
+      ]
+    }
+  }
+}
+```
+
+```mermaid
+sequenceDiagram
+    Form ->> CollectorFactory: collector(fields)
+    loop ForEach Field in Fields
+        CollectorFactory ->> Collector: create()
+        CollectorFactory ->> Collector: init(field)
+        Collector ->> Collector: populate the instance properties with field Json
+    end
+    CollectorFactory ->> Form: collectors
+```
+
+## How Collector populate default value
+
+The Collector populates the default value from the `formData` JSON:
+
+```json
+{
+  "formData": {
+    "value": {
+      "user.username": "",
+      "password": "",
+      "dropdown-field": "",
+      "combobox-field": [],
+      "radio-field": "",
+      "checkbox-field": []
+    }
+  }
+}
+```
+
+```mermaid
+sequenceDiagram
+    loop ForEach Collector in Collectors
+        Form ->> Collector: getKey()
+        alt key in formData
+            Form ->> Collector: init(formData[key])
+            Collector ->> Collector: "populate the default value with formData
+        end
+    end
+```
+
+## How Collector access ConnectorNode
+
+The Collector is self-contained and does not have access to the ConnectorNode by default. The Collector itself handle
+how to collect data from the user and how to validate the data. However, in some scenarios the Collector needs to access
+the ConnectorNode. For example, the Collector needs to access the root JSON to get the `passwordPolicy` to validate the
+password.
+
+To allow the Collector to access the ConnectorNode, the Collector can implement the `ConnectorNodeAware` interface.
+The `ConnectorNodeAware` interface provides the `connectorNode` property to access the ConnectorNode. After Collector is
+created, the ConnectorNode will be injected to the Collector.
+
+```kotlin
+class PasswordCollector : ContinueNodeAware {
+    override lateinit var continueNode: ContinueNode
+}
+```
+
+```mermaid
+sequenceDiagram
+    loop ForEach Collector in ContinueNode.collectors
+        alt Collector is ContinueNodeAware
+            CollectorFactory ->> Collector: setContinueNode(continueNode)
+        end
+    end
+```
+
+
+
+
```

Comment:
Agree with all comments above from Stoyan. Also, because it's a high level design, I believe it could easily fit in the README, avoiding the need of an additional file. However, I am ok with Stoyan's suggestion to include a link to this file. In this case, I would rename the file to something like COLLECTORS.

Reply:
I would keep prefer to keep it as 2 files for now, one is more for internal design which developer may not really care, and one for developer facing interface which aim for developer.